core/src/main/java/org/elasticsearch/index/query/QueryParser.java

```diff
@@ -38,7 +39,7 @@
      *            call
      * @return the new QueryBuilder
      */
-    QB fromXContent(QueryParseContext parseContext) throws IOException;
+    Optional<QB> fromXContent(QueryParseContext parseContext) throws IOException;
```

Comment:
Needed to change this signature so some QueryBuilders (like Boosting, ConstantScore, HasChild...) have the change to deffer resolving an empty Optional to something (thats what they also used to do before by returning 'null' to their caller in parse() (on 2.0) and now in doToQuery().